include/fi_tagged_rx.h

```diff
@@ -0,0 +1,156 @@
+ *
+ *         <optional: iovec's and other field(s)>
+ *         struct iovec iov[MAX_IOV_LIMIT];
+ *         void *desc[MAX_IOV_LIMIT];
+ * };
+ *
+ * Usage: 
+ *
+ * <name>_create - Create tagged-recv list, with a maximum of
+ * `num_rx' recv entries in freestack
+ *
+ * <name>_free - Frees the tagged-recv list
+ *
+ * <name>_entry_alloc - gets an entry from freestack. Caller can enqueue
+ * it to the posted/unexpected list as needed
+ *
+ * <name>_search_post - search for a matching recv entry with the given
+ * tag/addr in posted list
+ * 
+ * <name>_serch_unexp - search for a matching recv entry with the given
+ * tag/ignore/addr in unexpected list
+ */
+#define DECLARE_TAGGED_RX(entrytype, name)				\
+	DECLARE_FREESTACK(entrytype, name ## _fs)			\
+	struct name {							\
+		struct dlist_entry post_list;				\
+		struct dlist_entry unexp_list;				\
+		struct name ##_fs *fs;					\
+	};								\
+									\
+	static inline struct name *name ## _create(ssize_t num_rx)	\
+	{								\
+		struct name *list;					\
+		list = calloc(1, sizeof(struct name));			\
+		if (list) {						\
+			list->fs = name ## _fs_create(num_rx);		\
+				dlist_init(&list->post_list);		\
+				dlist_init(&list->unexp_list);		\
+		}							\
+		return list;						\
+	}								\
+									\
+	static inline void name ## _free(struct name *list)		\
+	{								\
+		name ## _fs_free(list->fs);				\
+		free(list);					        \
+	}								\
+									\
+	static inline entrytype *name ## _entry_alloc(struct name *list)\
+	{								\
+		int i;							\
+		entrytype *entry;					\
+		if (freestack_isempty(list->fs))			\
+			return NULL;					\
+		i = freestack_pop(list->fs);				\
+		entry = &list->fs->buf[i];				\
+		entry->fs_entry = i;					\
+		return entry;						\
+	}								\
+									\
+	static inline void name ## _entry_free(struct name *list,	\
+					       entrytype *entry)	\
+	{								\
+		freestack_push(list->fs, entry->fs_entry);		\
+	}								\
+									\
+	static inline entrytype *name ## _search_post(struct name *list,\
+						      uint64_t tag,	\
+						      fi_addr_t addr)	\
+	{								\
+		entrytype *entry;					\
+		struct dlist_entry *item;				\
+		dlist_foreach(&list->post_list, item) {			\
+			entry = container_of(item, entrytype, entry);	\
+			if ((entry->tmsg.tag & ~entry->tmsg.ignore) ==	\
+			    (tag & ~entry->tmsg.ignore) &&		\
+			    ((entry->tmsg.addr == FI_ADDR_UNSPEC) ||	\
+			     (addr == FI_ADDR_UNSPEC) ||		\
+			     (addr == entry->tmsg.addr)))		\
+				return entry;				\
+		}							\
+		return NULL;						\
+	}								\
+									\
+	static inline entrytype *name ## _search_unexp(struct name *list, \
+						       uint64_t tag,	\
+						       uint64_t ignore,	\
+						       fi_addr_t addr)	\
+	{								\
+		entrytype *entry;					\
+		struct dlist_entry *item;				\
+		dlist_foreach(&list->unexp_list, item) {		\
+			entry = container_of(item, entrytype, entry);	\
+			if ((entry->tmsg.tag & ~ignore) ==		\
+			    (tag & ~ignore) &&				\
```

Comment:
We may be better off using 'match' in place of 'ignore' both here and above.  There are 2 places where ignore bits are inverted.